src/tools/vx.8
.TH vx 8 User Manuals
.SH NAME
vx \- Virtual Context Setup
.SH SYNOPSIS
\fBvx -create \fIxid\fB [\fIlist\fB] [-- \fIprogram\fB \fIargs\fB*]

vx -migrate \fIxid\fB -- \fIprogram\fB \fIargs\fB*

vx -login \fIxid\fB [-- \fIprogram\fB \fIargs\fB*]

vx -set-bcaps \fIxid\fB \fIlist\fB

vx -set-ccaps \fIxid\fB \fIlist\fB

vx -set-flags \fIxid\fB \fIlist\fB

vx -set-limit \fIxid\fB \fItype\fB=\fImin\fB,\fIsoft\fB,\fIhard\fB*

vx -set-sched \fIxid\fB \fItype\fB=\fIvalue\fB*

vx -set-vhi \fIxid\fB \fItype\fB=\fIvalue\fB*

vx -get-bcaps \fIxid\fB

vx -get-ccaps \fIxid\fB

vx -get-flags \fIxid\fB

vx -get-limit \fIxid\fB \fItype\fB*

vx -get-vhi \fIxid\fB \fItype\fB*

vx -kill \fIxid\fB [\fIpid\fB \fIsig\fB]

vx -wait \fIxid\fB

vx -reset-limit \fIxid\fB
\f1
.SH DESCRIPTION
The purpose of this "Context" is to hide all processes outside of its scope, and prohibit any unwanted interaction between a process inside the context and a process belonging to another context.

This separation requires the extension of some existing data structures in order for them to become aware of contexts and to differentiate between identical uids used in different virtual servers.

It also requires the definition of a default context that is used when the host system is booted, and to work around the issues resulting from some false assumptions made by some user-space tools (like pstree) that the init process has to exist and to be running under id '1'.

To simplify administration, the Host Context isn't treated any differently than any other context as far as process isolation is concerned. To allow for process overview, a special Spectator context has been defined to peek at all processes at once.
.SH OPTIONS
.TP
\fB-create\f1
Create a new context with id \fIxid\f1 using optional flags in \fIlist\f1 and optionally run \fIprogram\f1 in the new context.
.TP
\fB-migrate\f1
Migrate to context \fIxid\f1 and execute \fIprogram\f1.
.TP
\fB-login\f1
Migrate to context \fIxid\f1 and execute \fIprogram\f1 using a new pseudo terminal.
.TP
\fB-set-bcaps\f1
Set system capabilities in \fIlist\f1 for context \fIxid\f1. See below for valid capabilities.
.TP
\fB-set-ccaps\f1
Set context capabilities in \fIlist\f1 for context \fIxid\f1. See below for valid capabilities.
.TP
\fB-set-flags\f1
Set context flags in \fIlist\f1 for context \fIxid\f1. See below for valid capabilities.
.TP
\fB-set-limit\f1
Set \fImin\f1, \fIsoft\f1 and \fIhard\f1 limit for \fItype\f1 and context \fIxid\f1. See below for valid limit types.
.TP
\fB-set-sched\f1
Set token bucket scheduler for context \fIxid\f1 using \fIvalue\f1. See below for valid scheduler types.
.TP
\fB-set-vhi\f1
Set virtual host information (UTS/uname) for context \fIxid\f1 using \fIvalue\f1. See below for valid VHI types.
.TP
\fB-get-bcaps\f1
Get system capabilities for context \fIxid\f1. See below for valid capabilities.
.TP
\fB-get-ccaps\f1
Get context capabilities for context \fIxid\f1. See below for valid capabilities.
.TP
\fB-get-flags\f1
Get context flags for context \fIxid\f1. See below for valid capabilities.
.TP
\fB-get-limit\f1
Get limit for \fItype\f1 and context \fIxid\f1. See below for valid limit types.
.TP
\fB-get-vhi\f1
Get virtual host information (UTS/uname) for context \fIxid\f1. See below for valid VHI types.
.TP
\fB-kill\f1
Kill process(es) in context \fIxid\f1 using optional process id \fIpid\f1 and signal number \fIsig\f1. Default signal number is 9.
.TP
\fB-wait\f1
Wait until context has died.
.TP
\fB-reset-limit\f1
Reset limit accounting for context \fIxid\f1.
.SH VALID SYSTEM CAPABILITIES
See \fBcapabilities(7)\f1. 
.SH VALID CONTEXT CAPABILITIES
SET_UTSNAME - allow \fBsetdomainname(2)\f1 and \fBsethostname(2)\f1

SET_RLIMIT - allow \fBsetrlimit(2)\f1

RAW_ICMP - allow raw ICMP sockets

SYSLOG - allow \fBsyslog(2)\f1

SECURE_MOUNT - allow secure \fBmount(2)\f1

SECURE_REMOUNT - allow secure remount

BINARY_MOUNT - allow binary/network mounts

QUOTA_CTL - allow quota ioctl
.SH VALID CONTEXT FLAGS
INFO_LOCK - Prohibit further context switches (legacy)

INFO_SCHED - Account all processes as one (legacy)

INFO_NPROC - Apply process limits to context (legacy)

INFO_PRIVATE - Context cannot be entered (legacy)

INFO_INIT - Show a fake init process

INFO_HIDE - Hide context information in task status

INFO_ULIMIT - Apply ulimits to context (legacy)

INFO_NSPACE - Use private namespace (legacy)

SCHED_HARD - Enable hard scheduler

SCHED_PRIO - Enable priority scheduler

SCHED_PAUSE - Pause context (unschedule)

VIRT_MEM - Virtualize memory information

VIRT_UPTIME - Virtualize uptime information

VIRT_CPU - Virtualize cpu usage information

VIRT_LOAD - Virtualize load average information

VIRT_TIME - Allow per guest time offsets

HIDE_MOUNT - Hide entries in /proc/$pid/mounts

HIDE_NETIF - Hide foreign network interfaces

STATE_SETUP - Context is in setup state

STATE_INIT - Context is in init state

SC_HELPER - Context state change helper

REBOOT_KILL - Kill all processes on \fBreboot(2)\f1

PERSISTENT - Make context persistent

FORK_RSS - Block fork when over RSS

PROLIFIC - Allow context to create new contexts

IGNEG_NICE - Ignore priority raise
.SH VALID LIMIT TYPES
CPU

FSIZE

DATA

STACK

CORE

RSS

NPROC

NOFILE

MEMLOCK

AS

LOCKS

NSOCK

OPENFD

ANON

SHMEM

SEMARY

NSEMS

DENTRY
.SH VALID SCHEDULER TYPES
FILL_RATE

FILL_RATE2

INTERVAL

INTERVAL2

TOKENS

TOKENS_MIN

TOKENS_MAX

PRIO_BIAS

IDLE_TIME

FORCE

CPU_ID

BUCKET_ID
.SH VALID VHI TYPES
CONTEXT

SYSNAME

NODENAME

RELEASE

VERSION

MACHINE

DOMAINNAME
.SH AUTHOR
vx was written by Benedikt Boehm <hollow@gentoo.org>

You will find the newest version of dx in the libvserver package at \fBhttp://linux-vserver.org\f1. There is also a mailing list, you will find information about it at Linux-VServer main site.
.SH BUGS
You should report them to the mailing list.
.SH SEE ALSO
\fBdx(8)\f1, \fBix(8)\f1, \fBns(8)\f1, \fBnx(8)\f1, \fBvr(8)\f1, 


NEWS
01 Aug 2005: First stable release of libvserver

I'm pleased to announce libvserver 0.3, the first release considered more or
less stable. This release supports the complete API of the Linux-VServer 2.0
kernel patches (libvserver.so) and provides additional helper functions such as
capability lists (libvserver-util.so). As of now, the public interface to
libvserver.so is not subject of change anymore.

An API doc is also available via doxygen, just use the provided Doxyfile or go
to http://dev.gentoo.org/~hollow/vserver/libvserver/doc/ for the latest API doc.

Have fun!


src/tools/vx.c
// Copyright 2006 Benedikt Böhm <hollow@gentoo.org>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the
// Free Software Foundation, Inc.,
// 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <inttypes.h>
#include <string.h>
#include <signal.h>
#include <lucid/bitmap.h>
#include <lucid/str.h>

#include "tools.h"
#include "vserver.h"

void do_vlogin(int argc, char *argv[], int ind);

static const char *rcsid = "$Id: vx.c 294 2006-07-09 08:46:15Z hollow $";

static
struct option long_opts[] = {
	COMMON_LONG_OPTS
	{ "create",      1, 0, 0x10 },
	{ "migrate",     1, 0, 0x11 },
	{ "set-bcaps",   1, 0, 0x12 },
	{ "set-ccaps",   1, 0, 0x13 },
	{ "set-flags",   1, 0, 0x14 },
	{ "set-limit",   1, 0, 0x15 },
	{ "set-sched",   1, 0, 0x16 },
	{ "set-vhi",     1, 0, 0x17 },
	{ "get-bcaps",   1, 0, 0x18 },
	{ "get-ccaps",   1, 0, 0x19 },
	{ "get-flags",   1, 0, 0x1A },
	{ "get-limit",   1, 0, 0x1B },
	{ "get-vhi",     1, 0, 0x1C },
	{ "limit-stat",  1, 0, 0x1D },
	{ "limit-reset", 1, 0, 0x1E },
	{ "sock-stat",   1, 0, 0x1F },
	{ "kill",        1, 0, 0x20 },
	{ "wait",        1, 0, 0x21 },
	{ NULL,          0, 0, 0 },
};

static
uint64_t str_to_rlim(char *str)
{
	if (str == NULL)
		return CRLIM_KEEP;
	
	if (strcmp(str, "inf") == 0)
		return CRLIM_INFINITY;
	
	if (strcmp(str, "keep") == 0)
		return CRLIM_KEEP;
	
	return atoi(str);
}

static
char *rlim_to_str(uint64_t lim)
{
	char *buf;
	
	if (lim == CRLIM_INFINITY)
		asprintf(&buf, "%s", "inf");
	else
		asprintf(&buf, "%" PRIu64, lim);
	
	return buf;
}

static inline
void usage(int rc)
{
	printf("Usage:\n\n"
	          "vx -create      <xid> [<list>] [-- <program> <args>*]\n"
	          "   -migrate     <xid> [<list>] -- <program> <args>*\n"
	          "   -login       <xid> [-- <program> <args>*]\n"
	          "   -set-bcaps   <xid> <list>\n"
	          "   -set-ccaps   <xid> <list>\n"
	          "   -set-flags   <xid> <list>\n"
	          "   -set-limit   <xid> <type>=<min>,<soft>,<hard>*\n"
	          "   -set-sched   <xid> <type>=<value>*\n"
	          "   -set-vhi     <xid> <type>=<value>*\n"
	          "   -get-bcaps   <xid>\n"
	          "   -get-ccaps   <xid>\n"
	          "   -get-flags   <xid>\n"
	          "   -get-limit   <xid> <type>*\n"
	          "   -limit-stat  <xid> <type>*\n"
	          "   -limit-reset <xid>\n"
	          "   -sock-stat   <xid> <type>*\n"
	          "   -get-vhi     <xid> <type>*\n"
	          "   -kill        <xid> [<pid> <sig>]\n"
	          "   -wait        <xid>\n"
	          );
	exit(rc);
}

int main(int argc, char *argv[])
{
	INIT_ARGV0
	
	int c, i;
	xid_t xid = 0;
	char *buf;
	uint64_t mask;
	
	/* syscall data */
	struct vx_create_flags cf = {
		.flags = 0,
	};
	
	struct vx_migrate_flags mf = {
		.flags = 0,
	};
	
	struct vx_bcaps bcaps = {
		.bcaps = ~(0ULL),
		.bmask = ~(0ULL),
	};
	
	struct vx_ccaps ccaps = {
		.ccaps =   0,
		.cmask =   0,
	};
	
	struct vx_flags flags = {
		.flags = 0,
		.mask  = 0,
	};
	
	struct vx_rlimit rlimit = {
		.id        = 0,
		.minimum   = CRLIM_KEEP,
		.softlimit = CRLIM_KEEP,
		.maximum   = CRLIM_KEEP,
	};
	
	struct vx_sched sched = {
		.set_mask   = 0,
		.fill_rate  = 0,
		.interval   = 0,
		.tokens     = 0,
		.tokens_min = 0,
		.tokens_max = 0,
		.prio_bias  = 0,
	};
	
	struct vx_vhi_name vhiname = {
		.field = 0,
		.name  = "",
	};
	
	struct vx_rlimit_stat rlimitstat = {
		.id      = 0,
		.hits    = 0,
		.value   = 0,
		.minimum = 0,
		.maximum = 0,
	};
	
	struct vx_sock_stat sockstat = {
		.field = 0,
		.count = { 0, 0, 0 },
		.total = { 0, 0, 0 },
	};
	
	struct vx_wait_result wait_result = {
		.reboot_cmd = 0,
		.exit_code  = 0,
	};
	
	struct vx_kill_opts kill_opts = {
		.pid = 0,
		.sig = SIGKILL,
	};
	
#define CASE_GOTO(ID, P) case ID: xid = atoi(optarg); goto P; break
	
	/* parse command line */
	while (GETOPT(c)) {
		switch (c) {
			COMMON_GETOPT_CASES
			
			CASE_GOTO(0x10, create);
			CASE_GOTO(0x11, migrate);
			CASE_GOTO(0x12, setbcaps);
			CASE_GOTO(0x13, setccaps);
			CASE_GOTO(0x14, setflags);
			CASE_GOTO(0x15, setlimit);
			CASE_GOTO(0x16, setsched);
			CASE_GOTO(0x17, setvhi);
			CASE_GOTO(0x18, getbcaps);
			CASE_GOTO(0x19, getccaps);
			CASE_GOTO(0x1A, getflags);
			CASE_GOTO(0x1B, getlimit);
			CASE_GOTO(0x1C, getvhi);
			CASE_GOTO(0x1D, limitstat);
			CASE_GOTO(0x1E, limitreset);
			CASE_GOTO(0x1F, sockstat);
			CASE_GOTO(0x20, kill);
			CASE_GOTO(0x21, wait);
			
			DEFAULT_GETOPT_CASES
		}
	}
	
#undef CASE_GOTO
	
	goto usage;
	
create:
	if (argc > optind && strcmp(argv[optind], "--") != 0) {
		if (flist64_from_str(argv[optind], cflags_list, &cf.flags, &mask, '~', ',') == -1)
			perr("flist64_from_str");
		
		if (vx_create(xid, &cf) == -1)
			perr("vx_create");
		
		if (argc > optind+2)
			execvp(argv[optind+2], argv+optind+2);
	}
	
	else {
		if (vx_create(xid, NULL) == -1)
			perr("vx_create");
		
		if (argc > optind+1)
			execvp(argv[optind+1], argv+optind+1);
	}
	
	goto out;

migrate:
	if (argc > optind && strcmp(argv[optind], "--") != 0) {
		if (flist64_from_str(argv[optind], mflags_list, &mf.flags, &mask, '~', ',') == -1)
			perr("flist64_from_str");
		
		if (vx_migrate(xid, &mf) == -1)
			perr("vx_migrate");
		
		if (argc > optind+2)
			execvp(argv[optind+2], argv+optind+2);
	}
	
	else {
		if (vx_migrate(xid, NULL) == -1)
			perr("vx_migrate");
		
		if (argc > optind+1)
			execvp(argv[optind+1], argv+optind+1);
	}
	
	goto out;
	
setbcaps:
	if (argc <= optind)
		goto usage;
	
	if (flist64_from_str(argv[optind], bcaps_list, &bcaps.bcaps, &bcaps.bmask, '~', ',') == -1)
		perr("flist64_from_str");
	
	if (vx_set_bcaps(xid, &bcaps) == -1)
		perr("vx_set_bcaps");
	
	goto out;
	
setccaps:
	if (argc <= optind)
		goto usage;
	
	if (flist64_from_str(argv[optind], ccaps_list, &ccaps.ccaps, &ccaps.cmask, '~', ',') == -1)
		perr("flist64_from_str");
	
	if (vx_set_ccaps(xid, &ccaps) == -1)
		perr("vx_set_ccaps");
	
	goto out;
	
setflags:
	if (argc <= optind)
		goto usage;
	
	if (flist64_from_str(argv[optind], cflags_list, &flags.flags, &flags.mask, '~', ',') == -1)
		perr("flist64_from_str");
	
	if (vx_set_flags(xid, &flags) == -1)
		perr("vx_set_flags");
	
	goto out;
	
setlimit:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		buf = strtok(argv[i], "=");
		
		if (buf == NULL)
			goto usage;
		
		if (!(rlimit.id = flist32_getval(rlimit_list, buf)))
			perr("flist32_getval");
		
		rlimit.id = v2i32(rlimit.id);
		
		if ((buf = strtok(NULL, ",")) == NULL)
			goto usage;
		else
			rlimit.minimum = str_to_rlim(buf);
		
		if ((buf = strtok(NULL, ",")) == NULL)
			goto usage;
		else
			rlimit.softlimit = str_to_rlim(buf);
		
		if ((buf = strtok(NULL, ",")) == NULL)
			goto usage;
		else
			rlimit.maximum = str_to_rlim(buf);
		
		if (vx_set_rlimit(xid, &rlimit) == -1)
			perr("vx_set_rlimit");
	}
	
	goto out;
	
setsched:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		buf = strtok(argv[i], "=");
		
		if (buf == NULL)
			goto usage;
		
		if (!(sched.set_mask = flist32_getval(sched_list, buf)))
			perr("flist32_getval");
		
		buf = strtok(NULL, "=");
		
		if (buf == NULL)
			goto usage;
		
		switch (sched.set_mask) {
			case VXSM_FILL_RATE:
				sched.fill_rate = atoi(buf);
				break;
			
			case VXSM_FILL_RATE2:
				sched.fill_rate = atoi(buf);
				break;
			
			case VXSM_INTERVAL:
				sched.interval = atoi(buf);
				break;
			
			case VXSM_INTERVAL2:
				sched.interval = atoi(buf);
				break;
			
			case VXSM_TOKENS:
				sched.tokens = atoi(buf);
				break;
			
			case VXSM_TOKENS_MIN:
				sched.tokens_min = atoi(buf);
				break;
			
			case VXSM_TOKENS_MAX:
				sched.tokens_max = atoi(buf);
				break;
			
			case VXSM_PRIO_BIAS:
				sched.prio_bias = atoi(buf);
				break;
			
			case VXSM_CPU_ID:
				sched.cpu_id = atoi(buf);
				break;
			
			case VXSM_BUCKET_ID:
				sched.bucket_id = atoi(buf);
				break;
		}
		
		if (vx_set_sched(xid, &sched) == -1)
			perr("vx_set_sched");
	}
	
	goto out;
	
setvhi:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		buf = strtok(argv[i], "=");
		
		if (buf == NULL)
			goto usage;
		
		if (!(vhiname.field = flist32_getval(vhiname_list, buf)))
			perr("flist32_getval");
		
		vhiname.field = v2i32(vhiname.field);
		
		buf = strtok(NULL, "=");
		
		if (buf == NULL)
			goto usage;
		
		strncpy(vhiname.name, buf, VHILEN-1);
		vhiname.name[VHILEN-1] = '\0';
		
		if (vx_set_vhi_name(xid, &vhiname) == -1)
			perr("vx_set_vhi_name");
	}
	
	goto out;
	
getbcaps:
	if (vx_get_bcaps(xid, &bcaps) == -1)
		perr("vx_get_bcaps");
	
	buf = flist64_to_str(bcaps_list, bcaps.bcaps, '\n');
	
	if (!str_isempty(buf))
		printf("%s\n", buf);
	
	free(buf);
	
	goto out;
	
getccaps:
	if (vx_get_ccaps(xid, &ccaps) == -1)
		perr("vx_get_ccaps");
	
	buf = flist64_to_str(ccaps_list, ccaps.ccaps, '\n');
	
	if (!str_isempty(buf))
		printf("%s\n", buf);
	
	free(buf);
	
	goto out;
	
getflags:
	if (vx_get_flags(xid, &flags) == -1)
		perr("vx_get_flags");
	
	buf = flist64_to_str(cflags_list, flags.flags, '\n');
	
	if (!str_isempty(buf))
		printf("%s\n", buf);
	
	free(buf);
	
	goto out;
	
getlimit:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		if (!(rlimit.id = flist32_getval(rlimit_list, argv[i])))
			perr("flist32_getval");
		
		rlimit.id = v2i32(rlimit.id);
		
		if (vx_get_rlimit(xid, &rlimit) == -1)
			perr("vx_get_rlimit");
		
		printf("%s=", argv[i]);
		
		buf = rlim_to_str(rlimit.minimum);
		printf("%s,", buf);
		free(buf);
		
		buf = rlim_to_str(rlimit.softlimit);
		printf("%s,", buf);
		free(buf);
		
		buf = rlim_to_str(rlimit.maximum);
		printf("%s\n", buf);
		free(buf);
	}
	
	goto out;
	
getvhi:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		if (!(vhiname.field = flist32_getval(vhiname_list, argv[i])))
			perr("flist32_getval");
		
		vhiname.field = v2i32(vhiname.field);
		
		if (vx_get_vhi_name(xid, &vhiname) == -1)
			perr("vx_get_vhi_name");
		
		printf("%s=%s\n", argv[i], vhiname.name);
	}
	
	goto out;
	
limitstat:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		if (!(rlimitstat.id = flist32_getval(rlimit_list, argv[i])))
			perr("flist32_getval");
		
		rlimitstat.id = v2i32(rlimitstat.id);
		
		if (vx_get_rlimit_stat(xid, &rlimitstat) == -1)
			perr("vx_get_rlimit_stat");
		
		printf("%s=%" PRIu32 ",%" PRIu64 ",%" PRIu64 ",%" PRIu64 "\n",
		       argv[i],
		       rlimitstat.hits, rlimitstat.minimum,
		       rlimitstat.value, rlimitstat.maximum);
	}
	
	goto out;
	
limitreset:
	if (vx_reset_rlimit(xid) == -1)
		perr("vx_reset_rlimit");
	
	goto out;
	
sockstat:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		if (!(sockstat.field = flist32_getval(sock_list, argv[i])))
			perr("flist32_getval");
		
		sockstat.field = v2i32(sockstat.field);
		
		if (vx_get_sock_stat(xid, &sockstat) == -1)
			perr("vx_get_sock_stat");
		
		printf("%s=%" PRIu32 "/%" PRIu64 ",%" PRIu32 "/%" PRIu64 ",%" PRIu32 "/%" PRIu64 "\n",
		       argv[i],
		       sockstat.count[0], sockstat.total[0],
		       sockstat.count[1], sockstat.total[1],
		       sockstat.count[2], sockstat.total[2]);
	}
	
	goto out;
	
wait:
	if (vx_wait(xid, &wait_result) == -1)
		perr("vx_wait");
	
	printf("reboot_cmd=%" PRIi32 "\n", wait_result.reboot_cmd);
	printf("exit_code=%" PRIi32 "\n", wait_result.exit_code);
	
	goto out;
	
kill:
	if (argc > optind+1) {
		kill_opts.pid = atoi(argv[optind]);
		kill_opts.sig = atoi(argv[optind+1]);
	}
	
	else if (argc > optind)
		goto usage;
	
	if (vx_kill(xid, &kill_opts) == -1)
		perr("vx_kill");
	
	goto out;
	
usage:
	usage(EXIT_FAILURE);

out:
	exit(EXIT_SUCCESS);
}
